src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs

```diff
@@ -340,23 +340,39 @@ private static BinderCheckTypeResult CheckTypeCore(IErrorContainer errorContaine
                 return new BinderCheckTypeResult() { Coercions = coercions };
             }
 
+            // Option sets are checked against typeWant for a match of the backing kind
+            //   Pre-V1: Ensure that Booleans only match bool valued option sets
+            //   V1: BackingKind matches (and Decimal can work with Number) and CanCoerceToBackingKind is true
+            //
+            // Checking against typeWant and not all alternateTypes (below) prevents, for example, Boolean option sets from being used in + operations,
+            // as it would require a coercion to number and then another to Boolean.  Limiting option set usage in this manner is a good thing.
+            // If the maker wants to allow Boolean + to work, they can use the Value function on the option set first.
+            //
+            // Option sets need not (and shouldn't be) listed in calls to CheckTypeCore.  For consistency,
+            // we always check option sets here against the backing types.
+            //
+            // Option sets can always coerce to string, V1 and pre-V1.
+            var nodeBackingKind = nodeType.OptionSetInfo?.BackingKind;
+            if (nodeBackingKind != null &&
+                (typeWant.Kind == DKind.String ||
+                ((!usePowerFxV1CompatibilityRules || nodeType.OptionSetInfo.CanCoerceToBackingKind) &&
+                   ((typeWant.Kind == DKind.Boolean && nodeBackingKind == DKind.Boolean) ||
+                    ((typeWant.Kind == DKind.Number || typeWant.Kind == DKind.Decimal) && nodeBackingKind == DKind.Number) ||
```

Comment:
>DKind.Number [](http://example.com/codeflow?start=109&length=12)

I'm assuming we only have backing kinds of String/Boolean/Number/Color. Do we have checks somewhere to enforce that?

Reply:
There is a Contracts.Assert at the top of EnumSymbol.cs, line 60. I assume Contracts.Assert is strong enough?  EnumStoreBuilder only works with EnumSymbols. What one would really like is a restriction at the IExternalOptionSet level, but I don't know how to do that with an interface definition. The way this logic is written in BinderUtils, it wouldn't match for another type, so more restrictive if an unexpected type were to enter.